Sources/SourceKitLSP/SourceKitServer+Scheme.swift

```diff
@@ -0,0 +1,78 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2020 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import LanguageServerProtocol
+import BuildServerProtocol
+
+extension SourceKitServer {
+
+  public func onSchemeChange(buildScheme: BuildScheme, workspace: Workspace) {
+    let transitiveDependenciesCallback: ([BuildTargetIdentifier]) -> Void = {
+      self.limitIndexVisibility(targets: $0, workspace: workspace)
+    }
+    collectTransitiveDependencies(
+      buildScheme: buildScheme,
+      workspace: workspace,
+      callback: transitiveDependenciesCallback
+    )
+  }
+  
+  func collectTransitiveDependencies(
+    buildScheme: BuildScheme,
+    workspace: Workspace,
+    callback: @escaping ([BuildTargetIdentifier]) -> Void
+  ) {
+    workspace.buildSystemManager.buildTargets { targetsResponse in
+      guard case let .success(targets) = targetsResponse else {
+        // TODO: Response error handling
+        callback([])
+        return
+      }
+      let targetMap = targets.reduce(into: [BuildTargetIdentifier: BuildTarget]()) {
+        $0[$1.id] = $1
+      }
+      var topLevelTargets = buildScheme.targets
+      var transitiveDeps = Set<BuildTargetIdentifier>()
+      while !topLevelTargets.isEmpty {
+        let targetID = topLevelTargets.popLast()!
```

Comment:
`removeLast`

Reply:
we need the element returned

Reply:
https://developer.apple.com/documentation/swift/array/2885764-removelast